### ISSUE TITLE
Missing Features for `gkernel`

### ISSUE THREAD

**Issue by water111** [water111/jak-project#49]
Tracking things we need before we can compile `gkernel` in the new compiler.
- [x] Defmethod
- [x] Method support
  - [x] Get type of boxed `object`
  - [x] Get method of `basic`
  - [x] Get method of unboxed
  - [x] Call method
  - [x] Define custom new methods
  - [x] `(new 'global ...)` support
- [x] Binteger support
- [x] Number conversions
- [x] Type coerce
- [x] `the` and `the-as`
  - [x] Alias
  - [x] Number conversion
- [x] `->` operator
  - [x] Access fields
    - [x] Deref fields
    - [x] Offset fields
  - [x] Set Places
    - [x] Constant Offset Places
    - [x] Other Places
  - [x] Deref pointers
  - [x] Array access
- [x] `&` operator
- [x] Pair Support
  - [x] `cons`
  - [x] static pairs (maybe not needed?)
  - [x] `car` and `cdr` and `Val`s for pairs
  - [x] `'()`
- [x] Inline asm
  - [x] `rlet`
- [x] Calling GOAL from C runs GOAL on a stack in GOAL's memory space.
- [x] `(new 'static ...)` for basics
- [x] `(new 'global 'inline-array ...)`
- [x] Support compound typespecs in `new`
- [x] `and` and `or`
- [x] Lots of stuff from `gcommon`
  - [ ] Check the order of evaluating things for everything in `gcommon`
- [x] Automatic Inspect Methods

Nice to have, but not essential
- [ ] `defenum` and enums
- [ ] `break`
- [ ] GOAL Crash Handler
- [ ] addr2line
- [x] Null pointer deref causes a crash
- [ ] Safety Features: https://github.com/water111/jak-project/issues/50
- [ ] Use a single `RegKind` type which also includes 32/128 bits and float/int for xmms.
